core/dive.c

```diff
@@ -704,54 +704,71 @@ void fixup_dc_duration(struct divecomputer *dc)
 
 /* Which cylinders had gas used? */
 #define SOME_GAS 5000
-static unsigned int get_cylinder_used(const struct dive *dive)
+static bool cylinder_used(const cylinder_t *cyl)
 {
-	int i;
-	unsigned int mask = 0;
+	int start_mbar, end_mbar;
 
-	for (i = 0; i < MAX_CYLINDERS; i++) {
-		const cylinder_t *cyl = dive->cylinder + i;
-		int start_mbar, end_mbar;
+	if (cylinder_nodata(cyl))
+		return false;
+	start_mbar = cyl->start.mbar ?: cyl->sample_start.mbar;
+	end_mbar = cyl->end.mbar ?: cyl->sample_end.mbar;
 
-		if (cylinder_nodata(cyl))
-			continue;
-		start_mbar = cyl->start.mbar ?: cyl->sample_start.mbar;
-		end_mbar = cyl->end.mbar ?: cyl->sample_end.mbar;
+	// More than 5 bar used? This matches statistics.c
+	// heuristics
+	return start_mbar > end_mbar + SOME_GAS;
+}
+
+/* Get list of used cylinders. Returns the number of used cylinders. */
+static int get_cylinder_used(const struct dive *dive, bool used[])
+{
+	int i, num = 0;
 
-		// More than 5 bar used? This matches statistics.c
-		// heuristics
-		if (start_mbar > end_mbar + SOME_GAS)
-			mask |= 1 << i;
+	for (i = 0; i < MAX_CYLINDERS; i++) {
+		used[i] = cylinder_used(dive->cylinder + i);
+		if (used[i])
+			num++;
 	}
-	return mask;
+	return num;
 }
 
-/* Which cylinders do we know usage about? */
-static unsigned int get_cylinder_known(const struct dive *dive, const struct divecomputer *dc)
+/* Are there any used cylinders which we do not know usage about? */
+static bool has_unknown_used_cylinders(const struct dive *dive, const struct divecomputer *dc,
+				       const bool used_cylinders[], int num)
 {
-	unsigned int mask = 0;
+	int idx;
 	const struct event *ev;
+	bool *used_and_unknown = malloc(MAX_CYLINDERS * sizeof(bool));
+	memcpy(used_and_unknown, used_cylinders, MAX_CYLINDERS * sizeof(bool));
 
 	/* We know about using the O2 cylinder in a CCR dive */
 	if (dc->divemode == CCR) {
 		int o2_cyl = get_cylinder_idx_by_use(dive, OXYGEN);
-		if (o2_cyl >= 0)
-			mask |= 1 << o2_cyl;
+		if (o2_cyl >= 0 && used_and_unknown[o2_cyl]) {
+			used_and_unknown[o2_cyl] = false;
+			num--;
+		}
 	}
 
 	/* We know about the explicit first cylinder (or first) */
-	mask |= 1 << explicit_first_cylinder(dive, dc);
+	idx = explicit_first_cylinder(dive, dc);
+	if (idx > 0 && used_and_unknown[idx]) {
+		used_and_unknown[idx] = false;
+		num--;
+	}
```

Comment:
This seems to change behavior. Why the "if (idx > 0" test?

It very much intentionally knew that 0 is a valid value for the case of not explicitly having a first cylinder - then cylinder zero _is_ the first cylinder. 

So that discard of 0 seems just wrong, and is an unexplained change from what the old code did.

Reply:
Good catch - the condition is wrong.

I reckon that I was thinking about the no-cylinders case and this was intended to read `idx >= 0`.

Nevertheless, neither does `explicit_first_cylinder` return `< 0` in the no-cylinder case, nor can we reach that place, since `per_cylinder_mean_depth()` returns early if there are no cylinders.